kroxylicious/src/main/java/io/kroxylicious/proxy/internal/DefaultFilterContext.java

```diff
@@ -171,12 +174,38 @@ public void forwardResponse(ApiMessage response) {
         if (!name.endsWith("ResponseData")) {
             throw new AssertionError("Attempt to use forwardResponse with a non-response: " + name);
         }
-        // TODO check we've not forwarded it already
+        if (decodedFrame instanceof RequestFrame) {
```

Comment:
Another option could be to introduce `Request` and `Response` flavours of the `KrpcFilterContext` with different implementations of `forwardResponse`. In that case you'd need to generate all the filters to take Krpc[Request|Response]FilterContext.

Reply:
I'm on the fence.  The fact that you the filter api writer persona would be prevented from `forwardRequest` from with a response handler, OTOH the simplicity of the current `KrpcFilterContext` object is appealing,

Reply:
Any input @SamBarker? Either way we can push it off to another PR

Reply:
+1 to a new PR. I've poked the idea of separate contexts previously so lean that way but don't have a strong use case in mind for them.